File: src/kanban/dispatch.rs
//! Makes the Kanban board a dispatchable source: implements the SDK
//! `DispatchProvider` over `kanban.db`, so the core `MCPDispatcher` can claim
//! `Ready` tasks, run a worker agent per task, and report the outcome back.

use std::sync::Arc;

use app_space_sdk::dispatch::{
    Capacity, DispatchProvider, McpServerSpec, Outcome, WorkItem, Workspace,
};
use async_trait::async_trait;

use crate::kanban::api::now;
use crate::kanban::db::Db;

/// How long a claim is valid before the dispatcher may reclaim it (seconds).
const LEASE_SECS: i64 = 15 * 60;

pub struct KanbanDispatchProvider {
    db: Arc<Db>,
    /// The MCP server a worker gets so it can `kanban_show`/`complete`/`block`
    /// its own card. In-process the daemon passes a native `Stdio` spec
    /// (`senclaw kanban-server`); standalone it's an `Http` spec at this app's URL.
    worker_mcp: McpServerSpec,
}

impl KanbanDispatchProvider {
    pub fn new(db: Arc<Db>, worker_mcp: McpServerSpec) -> Self {
        Self { db, worker_mcp }
    }

    /// Convenience for the standalone binary: reach the app's HTTP MCP by URL.
    pub fn http(db: Arc<Db>, base_url: impl Into<String>) -> Self {
        Self {
            db,
            worker_mcp: McpServerSpec::Http { name: "senclaw-kanban".into(), url: base_url.into() },
        }
    }

    fn guidance(card_id: i64) -> String {
        format!(
            "You are running Kanban task #{id}. First call the tool \
`mcp__senclaw-kanban__kanban_show` with card_id={id} to read its description, comments, and the \
summaries of the tasks it depends on. Do the work using your tools. You MUST finish with exactly \
one of: `mcp__senclaw-kanban__kanban_complete(card_id={id}, summary=…)` when the task is done, or \
`mcp__senclaw-kanban__kanban_block(card_id={id}, reason=…)` when you need human input or an \
external dependency. Never stop without calling one of them. For code-changing work, block with \
reason 'review-required: …' and attach details via `kanban_comment`.",
            id = card_id
        )
    }
}

#[async_trait]
impl DispatchProvider for KanbanDispatchProvider {
    async fn claim_ready(&self, cap: Capacity) -> anyhow::Result<Vec<WorkItem>> {
        let claimed = self.db.dispatch_claim(cap.total, cap.per_assignee, LEASE_SECS, now())?;
        let items = claimed
            .into_iter()
            .map(|c| {
                let mut prompt = c.title.clone();
                if !c.description.trim().is_empty() {
                    prompt.push_str("\n\n");
                    prompt.push_str(&c.description);
                }
                WorkItem {
                    id: c.id.to_string(),
                    assignee: c.assignee,
                    prompt,
                    guidance: Some(Self::guidance(c.id)),
                    mcp: vec![self.worker_mcp.clone()],
                    workspace: match &c.workspace_dir {
                        Some(dir) if !dir.trim().is_empty() => Workspace::Dir { path: dir.clone() },
                        _ => Workspace::Scratch,
                    },
                    depends_on: Vec::new(),
                    priority: prio_num(c.priority.as_deref()),
                    timeout_secs: None,
                }
            })
            .collect();
        Ok(items)
    }

    async fn heartbeat(&self, item_id: &str) -> anyhow::Result<()> {
        if let Ok(id) = item_id.parse::<i64>() {
            self.db.dispatch_heartbeat(id, LEASE_SECS, now())?;
        }
        Ok(())
    }

    async fn reclaim(&self) -> anyhow::Result<Vec<String>> {
        let ids = self.db.dispatch_reclaim(now())?;
        // Hermes-style promotion runs at the top of every dispatcher tick: todo
        // cards with all dependencies done move to Ready, where they get claimed.
        let promoted = self.db.dispatch_promote(now())?;
        if !promoted.is_empty() {
            tracing::info!(
                "[kanban] promoted {} task(s) todo→ready: {:?}",
                promoted.len(),
                promoted
            );
        }
        Ok(ids.into_iter().map(|i| i.to_string()).collect())
    }

    async fn finalize(&self, item_id: &str, outcome: Outcome) -> anyhow::Result<()> {
        let id: i64 = match item_id.parse() {
            Ok(v) => v,
            Err(_) => return Ok(()),
        };
        // Always release the claim first.
        self.db.clear_claim(id)?;

        let (_t, _d, _col, board_id) = match self.db.card_detail(id) {
            Ok(v) => v,
            Err(_) => return Ok(()), // card gone
        };
        let role = self.db.card_role(id)?.unwrap_or_default();

        // The worker's own kanban_complete/block is authoritative; finalize only
        // reconciles a card the worker left unresolved.
        match outcome {
            Outcome::Completed { summary, .. } => {
                if role != "done" {
                    if let Some(done) = self.db.column_by_role(board_id, "done")? {
                        self.db.move_card(id, done, 0, now())?;
                        let body = if summary.trim().is_empty() {
                            "auto-closed: agent returned without calling kanban_complete".to_string()
                        } else {
                            summary
                        };
                        let _ = self.db.add_comment(id, "dispatcher", &body, "complete", now());
                    }
                }
            }
            Outcome::Blocked { reason } => {
                if role != "blocked" {
                    if let Some(bl) = self.db.column_by_role(board_id, "blocked")? {
                        self.db.move_card(id, bl, 0, now())?;
                        let body = if reason.trim().is_empty() { "blocked".to_string() } else { reason };
                        let _ = self.db.add_comment(id, "dispatcher", &body, "block", now());
                    }
                }
            }
            Outcome::Failed { error } => {
                if role != "blocked" && role != "done" {
                    if let Some(bl) = self.db.column_by_role(board_id, "blocked")? {
                        self.db.move_card(id, bl, 0, now())?;
                        let _ = self.db.add_comment(id, "dispatcher", &format!("gave_up: {error}"), "block", now());
                    }
                }
            }
            Outcome::TimedOut => {
                if role != "blocked" && role != "done" {
                    if let Some(bl) = self.db.column_by_role(board_id, "blocked")? {
                        self.db.move_card(id, bl, 0, now())?;
                        let _ = self.db.add_comment(id, "dispatcher", "gave_up: run timed out", "block", now());
                    }
                }
            }
        }
        Ok(())
    }
}

fn prio_num(p: Option<&str>) -> i32 {
    match p {
        Some("urgent") => 3,
        Some("high") => 2,
        Some("medium") => 1,
        _ => 0,
    }
}

File: senclaw-sdk/senclaw-app-sdk/src/dispatch.ts
/**
 * Autonomous work dispatch — the app side of the contract.
 *
 * The daemon's `MCPDispatcher` engine can drive any app that exposes four
 * endpoints. Implement {@link DispatchProvider} over your own store, mount
 * {@link dispatchRouter} (Express) or {@link handleDispatch} (anything else),
 * and the engine will claim work from you, keep leases alive, recover items
 * whose worker died, and report terminal outcomes back.
 *
 * ```ts
 * import express from 'express';
 * import { dispatchRouter } from '@senclaw/space-sdk/dispatch';
 *
 * app.use('/api/dispatch', dispatchRouter(provider));
 * ```
 *
 * The wire shape is the Rust SDK's, field for field, because the same engine
 * parses both: snake_case JSON, `Outcome` tagged by `status`, `Workspace`
 * tagged by `kind`, `McpServerSpec` tagged by `transport`. Note `depends_on`
 * and `timeout_secs` — camelCase there is silently dropped by serde, which
 * surfaces as a dependency that never held rather than as an error.
 */

/** How many workers the dispatcher can spawn right now. */
export type Capacity = {
  /** Max items to claim across this source this tick. */
  total: number;
  /** Max concurrent items per assignee (worker lane). 0 = unlimited. */
  per_assignee: number;
};

/** Where a worker runs. */
export type Workspace =
  /** Fresh temp dir, discarded when the worker finishes. */
  | { kind: 'scratch' }
  /** A persistent absolute path. */
  | { kind: 'dir'; path: string }
  /** A git worktree, for coding tasks. */
  | { kind: 'worktree'; repo: string; branch?: string };

/**
 * An MCP server a worker needs.
 *
 * Prefer `stdio` — an `http` spec has to be bridged to stdio by the engine at
 * launch, which is one more process and one more failure mode.
 */
export type McpServerSpec =
  | { transport: 'stdio'; name: string; command: string; args?: string[]; env?: Record<string, string> }
  | { transport: 'http'; name: string; url: string };

/** A single dispatchable unit of work. */
export type WorkItem = {
  /** Source-scoped id, opaque to the engine. */
  id: string;
  /** The task to run — becomes the agent's user prompt. */
  prompt: string;
  /** Worker/persona to route to. Omitted = the source's default persona. */
  assignee?: string | null;
  /** Extra system-prompt block appended to the persona's own. */
  guidance?: string | null;
  /** MCP servers the worker gets, usually including this app's own tools. */
  mcp?: McpServerSpec[];
  /** Where the worker runs. Defaults to scratch. */
  workspace?: Workspace;
  /** Ids this item depends on. Already satisfied by the time you return it. */
  depends_on?: string[];
  /** Higher runs first. */
  priority?: number;
  /** Per-item run timeout. */
  timeout_secs?: number | null;
};

/** The terminal result of a worker run. */
export type Outcome =
  | { status: 'completed'; summary?: string; metadata?: unknown }
  | { status: 'blocked'; reason: string }
  | { status: 'failed'; error: string }
  | { status: 'timed_out' };

/** Implement over your own store, then mount it. */
export interface DispatchProvider {
  /**
   * Atomically claim up to `capacity` ready items.
   *
   * **Atomically** matters: the engine may poll again before these finish, and
   * an item handed out twice is run twice.
   */
  claimReady(capacity: Capacity): Promise<WorkItem[]> | WorkItem[];
  /** Extend the lease on an in-flight item. Optional — omit if you have no leases. */
  heartbeat?(itemId: string): Promise<void> | void;
  /** Return dead-worker/expired-lease items to ready; return their ids. */
  reclaim?(): Promise<string[]> | string[];
  /** Record a terminal outcome. Map it to your own states. */
  finalize(itemId: string, outcome: Outcome): Promise<void> | void;
}

/** The four actions, as the engine names them in the URL path. */
export type DispatchAction = 'poll' | 'heartbeat' | 'reclaim' | 'finalize';

/** What {@link handleDispatch} answers with: a status and a JSON body. */
export type DispatchResult = { status: number; body: unknown };

/**
 * Framework-agnostic core. Route `POST <prefix>/<action>` to this and send
 * back `status` + `body` however your server does that.
 *
 * Errors become `500 {error}` rather than propagating: the engine reads that
 * field and backs off, whereas an exception escaping into the HTTP layer
 * reaches it as a connection reset it cannot distinguish from a crash.
 */
export async function handleDispatch(
  provider: DispatchProvider,
  action: DispatchAction | string,
  body: unknown,
): Promise<DispatchResult> {
  const b = (body ?? {}) as Record<string, unknown>;
  const itemId = typeof b.item_id === 'string' ? b.item_id : '';
  try {
    switch (action) {
      case 'poll': {
        const c = (b.capacity ?? {}) as Partial<Capacity>;
        const items = await provider.claimReady({
          total: Number(c.total ?? 0),
          per_assignee: Number(c.per_assignee ?? 0),
        });
        return { status: 200, body: items };
      }
      case 'heartbeat':
        await provider.heartbeat?.(itemId);
        return { status: 200, body: { ok: true } };
      case 'reclaim':
        return { status: 200, body: (await provider.reclaim?.()) ?? [] };
      case 'finalize': {
        const outcome = (b.outcome ?? { status: 'failed', error: 'no outcome sent' }) as Outcome;
        await provider.finalize(itemId, outcome);
        return { status: 200, body: { ok: true } };
      }
      default:
        return { status: 404, body: { error: `unknown dispatch action: ${action}` } };
    }
  } catch (err) {
    return { status: 500, body: { error: err instanceof Error ? err.message : String(err) } };
  }
}

/**
 * An Express router with `/poll`, `/heartbeat`, `/reclaim`, `/finalize`.
 * Mount it at `/api/dispatch`.
 *
 * Express is already a dependency of the `/mcp` subpath, so this costs nothing
 * extra — but it is imported lazily so that importing this module from a
 * browser bundle (for the types) does not drag a server framework in with it.
 */
export async function dispatchRouter(provider: DispatchProvider) {
  const { Router, json } = await import('express');
  const router = Router();
  router.use(json({ limit: '4mb' }));
  for (const action of ['poll', 'heartbeat', 'reclaim', 'finalize'] as const) {
    router.post(`/${action}`, async (req, res) => {
      const { status, body } = await handleDispatch(provider, action, req.body);
      res.status(status).json(body);
    });
  }
  return router;
}

// -- constructors, so callers never hand-write a tagged union ---------------

export const workspace = {
  scratch: (): Workspace => ({ kind: 'scratch' }),
  dir: (path: string): Workspace => ({ kind: 'dir', path }),
  worktree: (repo: string, branch?: string): Workspace => ({ kind: 'worktree', repo, ...(branch ? { branch } : {}) }),
};

export const mcpServer = {
  stdio: (name: string, command: string, args: string[] = [], env: Record<string, string> = {}): McpServerSpec =>
    ({ transport: 'stdio', name, command, args, env }),
  http: (name: string, url: string): McpServerSpec => ({ transport: 'http', name, url }),
};

export const outcome = {
  completed: (summary = '', metadata: unknown = null): Outcome => ({ status: 'completed', summary, metadata }),
  blocked: (reason: string): Outcome => ({ status: 'blocked', reason }),
  failed: (error: string): Outcome => ({ status: 'failed', error }),
  timedOut: (): Outcome => ({ status: 'timed_out' }),
};
